libraries/ui/src/main/java/androidx/media3/ui/SubtitlePainter.java

```diff
@@ -470,4 +476,117 @@ private static boolean areCharSequencesEqual(
+   */
+  private static boolean containsRTL(@Nullable CharSequence input) {
+    if (input == null) {
+      return false;
+    }
+    int length = input.length();
+    for (int offset = 0; offset < length; ) {
+      int codePoint = Character.codePointAt(input, offset);
+      byte dir = Character.getDirectionality(codePoint);
+      if (dir == Character.DIRECTIONALITY_RIGHT_TO_LEFT ||
+          dir == Character.DIRECTIONALITY_RIGHT_TO_LEFT_ARABIC ||
+          dir == Character.DIRECTIONALITY_RIGHT_TO_LEFT_EMBEDDING ||
+          dir == Character.DIRECTIONALITY_RIGHT_TO_LEFT_OVERRIDE) {
+        return true;
+      }
+      offset += Character.charCount(codePoint);
+    }
+    return false;
+  }
+
+  /**
+   * Applies bidirectional (bidi) Unicode wrapping to each line of the given {@link CharSequence}.
+   * <p>
+   * This method ensures that text containing both left-to-right (LTR) and right-to-left (RTL)
+   * scripts is displayed correctly by wrapping each line using {@link BidiFormatter#unicodeWrap}.
+   * It forces LTR context for wrapping and preserves spans and line breaks.
+   *
+   * @param input the input text as a {@link CharSequence}, possibly containing mixed-direction text
+   * @return a {@link CharSequence} with each line wrapped for proper bidi rendering
+   */
+  public static CharSequence wrapBidiText(CharSequence input) {
+    BidiFormatter bidiFormatter = BidiFormatter.getInstance();
+
+    // Preserve span in the input text.
+    Spanned spannedInput = (Spanned)input;
+    Object[] spans = spannedInput.getSpans(0, input.length(), Object.class);
+
+    // Create arrays to track the start and end of each span after wrapping.
+    int[] spanStarts = new int[spans.length];
+    int[] spanEnds = new int[spans.length];
+    Arrays.fill(spanStarts, -1);
+    Arrays.fill(spanEnds, -1);
+
+    String[] lines = input.toString().split("\n");
+    List<String> wrappedLines = new ArrayList<>();
+
+    // Calculate the offset of each span after wrapping
+    int spanUpdate = 0;
+    int lineStart = 0;
+    for (String line : lines) {
+      // According to unicodeWrap documentation, this will either add 2 more characters or none
+      String wrappedLine = bidiFormatter.unicodeWrap(line, TextDirectionHeuristics.LTR, true);
+      int diff = wrappedLine.length() - line.length();
+      if (diff > 0) {
+        spanUpdate++;
+      }
+      for (int j = 0; j < spans.length; j++) {
+        // Each span start or end is updated only once
+        if ((spanStarts[j] < 0) &&
+            (spannedInput.getSpanStart(spans[j]) >= lineStart) &&
+            (spannedInput.getSpanStart(spans[j]) < lineStart + line.length())) {
+          spanStarts[j] = spanUpdate;
+        }
+        if ((spanEnds[j] < 0) &&
+            ((spannedInput.getSpanEnd(spans[j]) - 1) >= lineStart) &&
+            ((spannedInput.getSpanEnd(spans[j]) - 1) < lineStart + line.length())) {
+          spanEnds[j] = spanUpdate;
```

Comment:
I'm afraid I don't fully understand this code.

Is it intended to support spans that start on one line and end on a later line?

If I'm reading the code below on L579 correctly, if the start of span X is in `spanStarts[j]` then the end of the same span must be at the same index: `spanEnds[j]`.

But I'm not sure I understand how that invariant is enforced up here.

Would you consider pulling this `wrapBidiText` method out into a package-private class so we can add unit tests? That would give me more confidence that this works (and will keep working). See for example `SpannedToHtmlConverterTest.java`.